### DISCUSSION TITLE
Provide errorprone options in Java-based Gradle Plugin

### DISCUSSION THREAD

**Discussion by assayire** [tbroyer/gradle-errorprone-plugin#96] [Q&A] [answered]
I am trying to apply errorprone from a custom plugin of mine. In other words, my plugins sets up errorprone in the host project. I am writing my plugin in Java. What would be the equivalent of this in Java?

```groovy
tasks.named('compileJava') {
        options.errorprone {
            enabled = true
            allErrorsAsWarnings = false
            excludedPaths = '.*(/build/generated/).*'
            disable('InvalidParam') // False positives on records https://github.com/google/error-prone/issues/2321
        }
}
```

---

I tried this but says no such extension:

```java
project.getExtensions().configure("errorprone", (Action<ErrorProneOptions>) errorProneOptions -> {
        errorProneOptions.getEnabled().set(true);
        errorProneOptions.getAllErrorsAsWarnings().set(false);
        errorProneOptions.getExcludedPaths().set(".*(/build/generated/).*");
        errorProneOptions.disable("InvalidParam");
      });
```

I tried this but get build error saying invalid compiler option -XepAllErrorsAsWarnings, and so on) 😫 

```java
project.getTasks().withType(JavaCompile.class, javaCompile -> {
        javaCompile
          .getOptions()
          .getCompilerArgs()
          .addAll(List.of(
            "-XepAllErrorsAsWarnings:false",
            "-XepExcludedPaths:.*(/build/generated/).*",
            "-Xep:InvalidParam:OFF"
          ));
      });
```

Hope there is a way to apply/configure errorprone in Java.

Thanks 🙏 

Related but did not work:
- https://github.com/google/error-prone/issues/1226
- https://github.com/google/error-prone/issues/1136

**Comment by tbroyer** [comment, marked as answer]
The extension is on each `JavaCompile` task's `getOptions()`, which you have to cast to `ExtensionAware` (note also how I used withType/configureEach to [avoid eager task configuration](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:how_do_i_defer_configuration)):

```java
project.getTasks().withType(JavaCompile.class).configureEach(javaCompile ->
    ((ExtensionAware) javaCompile.getOptions()).getExtensions().configure(ErrorProneOptions.class, options -> {
        options.getEnabled().set(true); // This should be the default anyway
        options.getAllErrorsAsWarnings().set(false);
        options.getExcludedPaths().set(".*/build/generated/.*");
        options.disable("InvalidParam");
    });
```

**Reply by assayire** [reply]
Thanks a ton @tbroyer ... let me try that and report back.

**Reply by assayire** [reply]
Got it all working. Thanks a ton @tbroyer.
